File: Data Structures/8. Tree/Trie/trie.py
# python implementation of trie
class Node:
  # constructor
  def __init__(self):
    self.children = [None] * 26
    self.isEOW = False # True if is the end of the word

# trie class
class Trie:
  # constructor
  def __init__(self):
    self.root = self.getNode()

  # return a new trie node
  def getNode(self):
    return Node()

  # convert character to index
  def char_to_index(self, char):
    # using lowercase 'a' to 'z'
    return ord(char) - ord('a')
  
  # insert a word into the trie
  def insert(self, word):
    current = self.root
    for level in range(len(word)):
      index = self.char_to_index(word[level])

      # if current char not present
      if not current.children[index]:
        current.children[index] = self.getNode()
      current = current.children[index]
    
    # mark last node as leaf
    current.isEOW = True
  
  # search a given word
  def search(self, word):
    current = self.root
    for level in range(len(word)):
      index = self.char_to_index(word[level])
      if not current.children[index]:
        return False
      current = current.children[index]
    return current != None and current.isEOW

# ======================
# testing the Trie class
# ======================

words = ["the", "a", "there", "answer", "any", "by", "bye", "their"]
output = ["Not present in trie", "Present in tire"]
 
# Create a Trie object
t = Trie()
print("- Trie object created!")
print

# Construct trie
for word in words:
    t.insert(word)
print("- Words inserted!")
print

# Search for different keys
print("- Search for different words")
print("{} ---- {}".format("the",output[t.search("the")]))
print("{} ---- {}".format("these",output[t.search("these")]))
print("{} ---- {}".format("their",output[t.search("their")]))
print("{} ---- {}".format("thaw",output[t.search("thaw")]))

File: Data Structures/8. Tree/Trie/trie.cpp
// a trie data structure implementation
#include <iostream>
#include <vector> 
using namespace std;

// a trie node
struct Node {
  Node* children[26]; // an array of trie node
  bool isEOW;
};

// return a new trie node
Node* newNode() {
  Node* node = new Node;
  node -> isEOW = false;
  for (int i = 0; i < 26; i++)
    node -> children[i] = NULL;
  return node;
}

// get index for the current key charactor
int getIndex(char key) {
  // the difference of stored key and 'a' is the index
  return key - 'a';
}

// insert key into trie
void insert(Node* root, string word) {
  Node* current = root;

  // iterate the word and assign new nodes
  for (int i = 0; i < word.length(); i++) {
    int index = getIndex(word[i]);
    // if the key is not present, add node to given index
    if (!current -> children[index])
      current -> children[index] = newNode();
    current = current -> children[index]; // let current be the last created node
  }
  current -> isEOW = true; // mark last node as leaf
}

// search a given key in the trie
bool search(Node* root, string word) {
  Node* current = root;
  for (int i = 0; i < word.length(); i++) {
    int index = getIndex(word[i]);
    if (!current -> children[index]) // node not found
      return false;
    current = current -> children[index];
  }
  return (current != NULL && current -> isEOW);
}

int main() {
  // create a trie
  vector<string> keys{"the", "a", "there", "answer", "any", "by", "bye", "their"};
  Node* root = newNode();
  for (string word : keys)
    insert(root, word);
  cout << "- Trie created!" << endl;
  
  // search the trie
  cout << "\n- Searching for \"the\":" << endl;
  if (search(root, "the")) 
    cout << "\"the\" exists in the trie\n";
  else 
    cout << "\"the\" doesn't exist";
  cout << "\n- Searching for \"that\":" << endl;
  if (search(root, "that")) 
    cout << "\"that\" exists in the trie\n";
  else 
    cout << "\"that\" doesn't exist";
}